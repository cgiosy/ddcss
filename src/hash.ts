import xxh32 from "xxh32";

const encoder = new TextEncoder();

const hashCode = (str: string) => (
	(xxh32(encoder.encode(str)) >>> 1)
		.toString(36)
		.padStart(7, "_")
);

export default hashCode;
