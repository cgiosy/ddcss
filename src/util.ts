export const camelToKebab = (s: string) => s.replace(/[A-Z]/g, (x) => `-${x.toLowerCase()}`);
