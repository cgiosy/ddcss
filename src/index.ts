import { camelToKebab } from "./util";
import hashCode from "./hash";

const tokenSymbol = Symbol("$$css.token");

type Falsy = false | 0 | 0n | "" | null | undefined | void;

type Digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";
type Uppercase = "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" | "J" | "K" | "L" | "M"
				| "N" | "O" | "P" | "Q" | "R" | "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z";
type Lowercase = "a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m"
				| "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z";
type Char = "_" | Digit | Uppercase | Lowercase;

type Macro = {
	fn: MacroFn,
	table: MacroTable | null,
};
type MacroFn = (arg: any) => CSSObject;
type TokenFn = (key: string, property: string) => string | Falsy;
type MacroTable = {
	[K: string]: Macro;
	[tokenSymbol]: TokenFn;
};

type Callback = () => void;
type TickFn = (callback: Callback) => unknown;
type FlushFn = (textContent: string) => string | void;
type FilterFn = (hash: string, obj: CSSObject) => boolean;
type FilterGenerator = () => FilterFn;

export type CSSObject = {
	[K: string]: any;
	[K: `@${string}` | `${string}&${string}`]: CSSObject;
	[K: `$$${Char}${string}`]: MacroFn;
	[K: `$${Char}${string}`]: string | number;
	$$?: TokenFn;
};

const isProperty = (key: string): key is `$${Char}${string}` => /^\$?[0-9A-Za-z_]+$/.test(key);
const isMacroKey = (key: string): key is `$$${Char}${string}` => key[0] === "$" && key[1] === "$";
const isAtRule = (key: string): key is `@${string}` => key[0] === "@";

const keyToProp = (key: string) => (
	key[0] === "$"
		? `--${camelToKebab(key.slice(1))}`
		: camelToKebab(key)
);

const tokenFallback = (key: string) => (
	key[0] === "$"
		? `var(--${camelToKebab(key.slice(1))})`
		: key
);

const copy = (obj: any) => Object.assign(Object.create(Object.getPrototypeOf(obj)), obj);

const chain = <T extends unknown[], T1, T2>(fn: (...args: T) => T1, next: (...args: T) => T2) => (
	(...args: T): Exclude<T1, Falsy> | T2 => (fn(...args) || next(...args)) as any
);

const initialMacros: MacroTable = Object.assign(
	Object.create(null),
	{ [tokenSymbol]: tokenFallback },
);

const stringify = (
	obj: CSSObject,
	parent: string,
	macros: MacroTable,
	outMacros?: MacroTable,
) => {
	const { classBody, outsideCss } = _stringify(obj, parent, macros, outMacros);
	return classBody === ""
		? outsideCss
		: `${parent}{${classBody}}${outsideCss}`;
};

const _stringify = (
	obj: CSSObject,
	parent: string,
	inMacros: MacroTable,
	outMacros?: MacroTable,
) => {
	const macros: MacroTable = Object.create(inMacros);
	let classBody = "";
	let outsideCss = "";

	for (const key of Object.keys(obj)) {
		let value = obj[key];
		if (isMacroKey(key)) {
			if (key.length === 2) {
				macros[tokenSymbol] = chain(value, macros[tokenSymbol]);
				continue;
			}

			const macroKey = key.slice(2);
			const guard = macroKey[0]!.toUpperCase() !== macroKey[0];
			const macro: Macro = {
				fn: value,
				table: guard ? copy(macros) : null,
			};
			macros[macroKey] = macro;
		} else if (key in macros) {
			const { fn, table } = macros[key]!;
			const result = _stringify(fn(value), parent, table ?? copy(macros), macros);
			classBody += result.classBody;
			outsideCss += result.outsideCss;
		} else if (isProperty(key)) {
			value = value.toString();
			const macro = macros[tokenSymbol];
			const prop = keyToProp(key);
			const len = value.length | 0;
			let body = "";
			let cur = 0;
			let last = 0;
			let quote = 0;
			while (cur < len) {
				const c = value.charCodeAt(cur);
				const next = cur + 1 | 0;
				if (quote === 0) {
					// [^0-9A-Za-z$_\\-]
					if (!(
						48 <= c && c <= 57
						|| 65 <= c && c <= 90
						|| 97 <= c && c <= 122
						|| c === 36
						|| c === 45
						|| c === 92
						|| c === 95
					)) {
						if (last < cur) {
							body += macro(value.slice(last, cur), key);
							last = cur;
						}

						// [^'"]
						if (c !== 34 && c !== 39) {
							body += String.fromCharCode(c);
							last = next;
						} else quote = c;
					}
				} else if (quote === c) {
					quote = 0;
					body += macro(value.slice(last, next), key);
					last = next;
				}
				cur = next;
			}
			if (last < len) body += macro(value.slice(last), key);
			classBody += `${prop}:${body};`;
		} else if (isAtRule(key)) {
			const body = stringify(value, parent, macros);
			if (body !== "") outsideCss += `${key}{${body}}`;
		} else {
			const selectors = (
				parent
					.split(",")
					.map((parentSelector) => {
						const selector = parentSelector.trim();
						return key.replace(/\\?&/g, (match) => (
							match[0] === "\\"
								? match
								: selector
						));
					})
					.join(",")
			);
			outsideCss += stringify(value, selectors, macros);
		}
	}
	if (outMacros !== undefined) Object.assign(outMacros, macros);

	return {
		classBody,
		outsideCss,
	};
};

const addToHead = (textContent: string) => {
	const style = document.createElement("style");
	style.textContent = textContent;
	document.head.appendChild(style);
};

const basicFilterGenerator = () => {
	const filterSet = new Set();
	return (hash: string) => {
		if (filterSet.has(hash)) return false;
		filterSet.add(hash);
		return true;
	};
};

const $$css = (globalObj: CSSObject | CSSObject[] = {}, {
	root = ":root",
	tick = (queueMicrotask || setTimeout) as TickFn,
	flush = addToHead as FlushFn,
	filter = basicFilterGenerator as FilterGenerator,
} = {}) => {
	const checkFilter = filter();
	const macros = Object.create(initialMacros);
	let textContent = "";
	let flushing = false;
	const tickFlush = (str = "") => {
		textContent += str;
		if (textContent !== "" && !flushing) {
			flushing = true;
			tick(tickFlush2);
		}
	};
	const tickFlush2 = () => {
		flushing = false;
		textContent = flush(textContent) || "";
		tickFlush();
	};
	for (const obj of [globalObj].flat())
		tickFlush(stringify(obj, root, copy(macros), macros));

	const $css = (obj: CSSObject, className: string) => stringify(obj, className, macros);
	const css = (obj: CSSObject, className?: string) => {
		if (className !== undefined) {
			const result = $css(obj, className);
			const hash = hashCode(result);
			if (checkFilter(hash, obj)) tickFlush(result);
			return className;
		}

		const result = $css(obj, ".$&");
		const hash = hashCode(result);
		if (checkFilter(hash, obj)) tickFlush(result.replace(/\.\$&/g, `.${hash}`));
		return hash;
	};
	return {
		$css,
		css,
	};
};

export default $$css;
