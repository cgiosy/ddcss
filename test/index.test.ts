import { describe, test } from "vitest";

import $$css, { CSSObject } from "../src";

const ddcss = () => {
	let output = "";
	const config = {
		tick: (callback: () => void) => callback(),
		flush: (text: string) => {
			output += text;
		},
	};
	return {
		$$css: (obj?: CSSObject, userConfig?: object) => $$css(obj, {
			...config,
			...userConfig,
		}),
		getOutput: () => output,
	};
};

describe("basic", () => {
	test("empty args", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css();
		css({});
		expect(getOutput()).toMatchSnapshot();
	});

	test("basic props and children", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({ $16dp: "1rem" });
		css({
			"& > *": {
				fontSize: "$16dp",
				color: "red",
			},
		});
		expect(getOutput()).toMatchSnapshot();
	});

	test("basic macros", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$font: (value) => ({
				fontSize: `${value}px`,
				fontFamily: "serif",
			}),
		});
		css({
			"& > *": {
				font: 32,
				color: "red",
			},
		});
		expect(getOutput()).toMatchSnapshot();
	});

	test("basic multiple css with macros", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$font: (value) => ({
				fontSize: `${value}px`,
				fontFamily: "serif",
			}),
		});
		css({
			$$font2: (value) => ({ font: value * 2 }),
			font2: 16,
		});
		css({
			$$font2: (value) => ({ font: value * 3 }),
			font2: 16,
		});
		expect(getOutput()).toMatchSnapshot();
	});

	test("basic duplicate filtering", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$font: (value) => ({
				fontSize: `${value}px`,
				fontFamily: "serif",
			}),
		});
		css({ font: 16 });
		css({ font: 16 });
		expect(getOutput()).toMatchSnapshot();
	});

	test("basic macro variable", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$: (key) => key.match(/^\d+dp$/) && `${Number(key.slice(0, -2)) / 16}rem`,
			fontSize: "16dp",
		});
		css({ fontSize: 14 });
		css({ fontSize: 18 });
		expect(getOutput()).toMatchSnapshot();
	});
});

describe.concurrent("macros", () => {
	test("recursive-like macros", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$userSelect: (value) => ({
				WebkitUserDrag: value,
				WebkitUserSelect: value,
				userSelect: value,
			}),
		});
		css({ userSelect: "none" });
		expect(getOutput()).toMatchSnapshot();
	});

	test("nested macros", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$f: (value) => ({ $x: `${value}f` }),
			$$g: (value) => ({ $y: `${value}g` }),
			"& > *": {
				$$f: (value) => ({ f: `${value}F` }),
				$$g: (value) => ({ g: `${value}G` }),
				f: ">",
				g: ">",
			},
			f: "&",
			g: "&",
		});
		css({
			f: "^",
			g: "^",
		});
		expect(getOutput()).toMatchSnapshot();
	});

	test("complex macros", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		$$css({
			$$f: (a) => ({ $$g: (b) => ({ $$h: (c) => ({ $x: a * 100 + b * 10 + c }) }) }),
			"& > *": {
				"& > *": {
					f: 1,
					g: 2,
					h: 3,
				},
				f: 4,
				g: 5,
				h: 6,
			},
			h: 9,
		});
		expect(getOutput()).toMatchSnapshot();
	});

	test("complex macro variables", ({ expect }) => {
		const { $$css, getOutput } = ddcss();
		const { css } = $$css({
			$$dp: (value) => ({
				fontSize: value > 0 ? `$$${value}dp` : "1rem",
				$$: (key) => key.match(/^\d+dp$/) && `${Number(key.slice(0, -2)) / 16}rem`,
			}),
			dp: 0,

			$$f: (value) => ({ dp: value * 2 }),
			$$g: (value) => ({ dp: value * 3 }),
		});
		css({
			dp: 16,
			"& > *": {
				$$f: (value) => ({ f: value * 4 }),
				$$g: (value) => ({ g: value * 5 }),
				f: 1,
				g: 2,
			},
			f: 1,
			g: 2,
		});
		css({
			dp: 14,
			"& > *": {
				$$f: (value) => ({ f: value * 4 }),
				$$g: (value) => ({ g: value * 5 }),
				f: 1,
				g: 2,
			},
			f: 1,
			g: 2,
		});
		expect(getOutput()).toMatchSnapshot();
	});
});
