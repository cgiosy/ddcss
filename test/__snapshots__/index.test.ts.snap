// Vitest Snapshot v1

exports[`basic > basic duplicate filtering 1`] = `"._djeiow{font-size:16px;font-family:serif;}"`;

exports[`basic > basic macro variable 1`] = `":root{font-size:16dp;}._kln0i3{font-size:14;}._jqlkvk{font-size:18;}"`;

exports[`basic > basic macros 1`] = `".__w9ras > *{font-size:32px;font-family:serif;color:red;}"`;

exports[`basic > basic multiple css with macros 1`] = `"._rvbeky{font-size:32px;font-family:serif;}._nls01e{font-size:48px;font-family:serif;}"`;

exports[`basic > basic props and children 1`] = `":root{--16dp:1rem;}._hrwqpr > *{font-size:var(--16dp);color:red;}"`;

exports[`basic > empty args 1`] = `""`;

exports[`macros > complex macro variables 1`] = `":root{font-size:1rem;}._54trbc{font-size:var(--16dp);font-size:var(--2dp);font-size:var(--6dp);}._54trbc > *{font-size:var(--8dp);font-size:var(--30dp);}._epi98u{font-size:var(--14dp);font-size:var(--2dp);font-size:var(--6dp);}._epi98u > *{font-size:var(--8dp);font-size:var(--30dp);}"`;

exports[`macros > complex macros 1`] = `":root{h:9;}:root > *{--x:456;}:root > * > *{--x:123;}"`;

exports[`macros > nested macros 1`] = `":root{--x:&f;--y:&g;}:root > *{--x:>Ff;--y:>Gg;}._jzlwen{--x:^f;--y:^g;}"`;

exports[`macros > recursive-like macros 1`] = `"._8d3u5t{-webkit-user-drag:none;-webkit-user-select:none;user-select:none;}"`;
